src/profile/hooks/useProfileLoadTracker.ts

```diff
@@ -0,0 +1,139 @@
+import { useCallback, useContext, useEffect, useRef } from 'react';
+
+import { ProfileContext, ProfileContextData } from '../context/ProfileContext';
+import parseGraphQLError from '../helpers/parseGraphQLError';
+import { QueryResult } from './useProfileQuery';
+
+const STATUS_FAILED = -1;
+const STATUS_WAITING = 0;
+const STATUS_SHOULD_LOAD = 1;
+const STATUS_LOADING = 2;
+const STATUS_SUCCESS = 3;
+
+export type Status =
+  | typeof STATUS_WAITING
+  | typeof STATUS_LOADING
+  | typeof STATUS_FAILED
+  | typeof STATUS_SUCCESS
+  | typeof STATUS_SHOULD_LOAD;
+
+export type useProfileLoaderHookReturnType = {
+  isProfileLoadComplete: () => boolean;
+  hasExistingProfile: () => boolean;
+  didProfileLoadFail: () => boolean;
+  reloadProfile: () => Promise<QueryResult>;
+  getStatus: () => Status;
+  getProfile: () => ReturnType<ProfileContextData['getProfile']>;
+};
+
+export function shouldLoad(currentStatus?: Status): boolean {
+  return currentStatus === STATUS_SHOULD_LOAD;
+}
+
+export function getShouldLoadStatus(): Status {
+  return STATUS_SHOULD_LOAD;
+}
+
+export function isLoading(currentStatus?: Status): boolean {
+  return currentStatus === STATUS_LOADING;
+}
+
+export function getLoadingStatus(): Status {
+  return STATUS_LOADING;
+}
+
+export function didLoadFail(currentStatus?: Status): boolean {
+  return currentStatus === STATUS_FAILED;
+}
+
+export function getLoadFailedStatus(): Status {
+  return STATUS_FAILED;
+}
+
+export function isLoadComplete(currentStatus?: Status): boolean {
+  return didLoadFail(currentStatus) || isLoadedSuccessfully(currentStatus);
+}
+
+export function getLoadSuccessStatus(): Status {
+  return STATUS_SUCCESS;
+}
+
+export function isLoadedSuccessfully(currentStatus?: Status): boolean {
+  return currentStatus === STATUS_SUCCESS;
+}
+
+function canStartLoading(currentStatus?: Status): boolean {
+  return !isLoading(currentStatus) && !isLoadComplete(currentStatus);
+}
+
+export function getInitialStatus(): Status {
+  return STATUS_WAITING;
+}
+
+export function resolveChangedStatus(
+  currentStatus: Status,
+  profileContext: ProfileContextData
+): Status | undefined {
+  if (isLoadComplete(currentStatus)) {
+    return undefined;
+  }
+
+  if (
+    profileContext.error &&
+    !parseGraphQLError(profileContext.error).isAllowedError
+  ) {
+    return getLoadFailedStatus();
+  }
+
+  if (profileContext.data && !isLoadedSuccessfully(currentStatus)) {
+    return getLoadSuccessStatus();
+  }
+
+  if (canStartLoading(currentStatus) && !shouldLoad(currentStatus)) {
+    return getShouldLoadStatus();
+  }
+
+  return undefined;
+}
+
+export function useProfileLoadTracker(): useProfileLoaderHookReturnType {
+  const profileContext = useContext(ProfileContext);
+  const valueRef = useRef<Status>(STATUS_WAITING);
```

Comment:
The value's name should be something more explanatory. Perhaps something like `currentStatus`.

Reply:
Changed to "statusRef", because there is also "updateStatus", not "updateCurrentStatus" and currentStatusRef.current sounds weird :)